clippy_lints/src/loops/explicit_counter_loop.rs

```diff
@@ -34,24 +34,29 @@ pub(super) fn check<'tcx>(
                 if let Some((name, ty, initializer)) = initialize_visitor.get_result();
                 if is_integer_const(cx, initializer, 0);
                 then {
-                    let mut applicability = Applicability::MachineApplicable;
+                    let mut applicability = Applicability::MaybeIncorrect;
 
                     let int_name = match ty.map(Ty::kind) {
                         // usize or inferred
                         Some(ty::Uint(UintTy::Usize)) | None => {
-                            span_lint_and_sugg(
+                            span_lint_and_then(
                                 cx,
                                 EXPLICIT_COUNTER_LOOP,
                                 expr.span.with_hi(arg.span.hi()),
                                 &format!("the variable `{}` is used as a loop counter", name),
-                                "consider using",
-                                format!(
-                                    "for ({}, {}) in {}.enumerate()",
-                                    name,
-                                    snippet_with_applicability(cx, pat.span, "item", &mut applicability),
-                                    make_iterator_snippet(cx, arg, &mut applicability),
-                                ),
-                                applicability,
+                                |diag| {
+                                    diag.span_suggestion(
+                                        expr.span.with_hi(arg.span.hi()),
```

Comment:
This should be done once before the `span_lint` call.

Reply:
I commited https://github.com/rust-lang/rust-clippy/pull/9149/commits/134d0dee41846d1c0867fa3745b8de25b52ae7af